src/charm.py

```diff
@@ -142,38 +145,73 @@ def __init__(self, *args):
     def _on_install(self, event: InstallEvent):
         """Handle the install charm event."""
         if not self._container.can_connect():
-            self.unit.status = WaitingStatus("Waiting to be able to connect to vault unit")
             event.defer()
             return
         self._delete_vault_data()
 
+    def _on_collect_status(self, event: CollectStatusEvent):
+        """Handle the collect status event."""
+        if (
+            self._tls_certificates_pki_relation_created()
+            and not self._common_name_config_is_valid()
+        ):
+            event.add_status(
+                BlockedStatus(
+                    "Common name is not set in the charm config, cannot configure PKI secrets engine"
+                )
+            )
+            return
+        if not self._container.can_connect():
+            event.add_status(WaitingStatus("Waiting to be able to connect to vault unit"))
+            return
+        if not self._is_peer_relation_created():
+            event.add_status(WaitingStatus("Waiting for peer relation"))
+            return
+        if not self._bind_address or not self._ingress_address:
+            event.add_status(
+                WaitingStatus("Waiting for bind and ingress addresses to be available")
+            )
+            return
+        if not self.unit.is_leader() and len(self._other_peer_node_api_addresses()) == 0:
+            event.add_status(WaitingStatus("Waiting for other units to provide their addresses"))
+            return
+        if not self.unit.is_leader() and not self.tls.ca_certificate_is_saved():
+            event.add_status(WaitingStatus("Waiting for CA certificate"))
+            return
+        if not self.unit.is_leader() and not self._initialization_secret_set_in_peer_relation():
+            event.add_status(WaitingStatus("Waiting for initialization secret"))
+            return
+        if not self.tls.tls_file_pushed_to_workload(File.CA):
+            event.add_status(WaitingStatus("Waiting for TLS certificates"))
+            return
+        if not self.tls.tls_file_available_in_charm(File.CA):
+            event.add_status(WaitingStatus("Waiting for CA certificate"))
+            return
```

Comment:
I'm not sure I understand the difference between the two.

Reply:
On storage detaching event, it is possible for one of those conditions to be true and not the other (i.e. the tls file can be in the workload but not available in the charm), hence the need for the 2. I improved the content of the status message to make it a bit clearer.